Sources/Workspace/Workspace.swift

```diff
@@ -1975,6 +2034,74 @@ extension Workspace {
 
         return result.map{ $0 }
     }
+    
+    private func extract(_ artifacts: [LocalArchivedArtifact], diagnostics: DiagnosticsEngine) -> [ManagedArtifact] {
```

Comment:
The DRY principle is not completely followed here since the function copies a huge chunk of the extraction logic from `func download(_ artifacts: [RemoteArtifact] ...) -> [ManagedArtifact]`, I went this way to minimize the amount of refactoring in the PR, but if it is critical I could introduce a more generic function e.g. `func extract(_ artifacts: [UnmanagedArtifact] ...) -> [ManagedArtifacts]`, a price for that would be having more complicated logic and a few additional types, or come up with something else.

Reply:
I think this is fine for now, once the PR is more mature we can see if we can make this DRYer

Reply:
Resolving this one, since the discussion is continued [in the comments](https://github.com/apple/swift-package-manager/pull/3746#issuecomment-934372019).